Upgrade sqlserver driver to version 12.8.1

lukmanulhakkeem opened the pull request:
## Description
<!---Describe your changes in detail-->

NTLM authentication is not supported in the current driver. We will get below error if we specify NTLM in the connection URL
`integratedSecurity=true;authenticationScheme=NTLM;`

![image](https://github.com/user-attachments/assets/e95a311b-c1d3-4d1a-983b-cf070624bfee)

We need to upgrade the driver to get this support

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

SQL Server Connector Changes
* Upgrade sqlserver driver to version 12.8.1, to include NTLM authentication support. See :ref:`connector/sqlserver:authentication`.
```

lukmanulhakkeem commented on pom.xml:
```diff
@@ -1385,7 +1385,7 @@
                 <groupId>com.microsoft.sqlserver</groupId>
                 <artifactId>mssql-jdbc</artifactId>
-                <version>7.0.0.jre8</version>
+                <version>8.2.1.jre8</version>
```
My bad, actually i meant for 12.8.1. I will change it

lukmanulhakkeem replied:
@imjalpreet when are we planning for the java upgrade?

ZacBlanco replied:
We hope to be running mostly java 17 bytecode by end of H1/beginning of H2. For now, we will be on java 8

imjalpreet replied:
@lukmanulhakkeem can you please look into the CI failures?

lukmanulhakkeem replied:
@ZacBlanco So this driver needs to be updated with the java 17 upgrade right? How can we ensure that? Is there any parent ticket that track all the dependencies for the java 17 upgrade?

pratyakshsharma replied:
@lukmanulhakkeem you might find this helpful - https://github.com/orgs/prestodb/projects/31/views/1

lukmanulhakkeem commented:
@imjalpreet ci failures are fixed. could you review it again?

pratyakshsharma reviewed (dismissed):
thank you for the fix, looks good.

steveburnett commented:
Consider if we should add a mention in the [SQL Server Connector](https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/connector/sqlserver.rst) doc that NTLM is supported. There isn't a mention in the doc of the current absence of NTLM as a limitation.

imjalpreet reviewed (changes requested):
@lukmanulhakkeem Thanks for the fix, in addition to what Steve suggested can we also update the documentation to highlight that the default value of `encrypt` is `true` after the driver upgrade?

lukmanulhakkeem commented:
@steveburnett @imjalpreet I will update

lukmanulhakkeem commented:
> @lukmanulhakkeem Thanks for the fix, in addition to what Steve suggested can we also update the documentation to highlight that the default value of `encrypt` is `true` after the driver upgrade?


This default value is already updated in the documentation by mistake. So I can update only the NTLM auth support

<img width="1079" alt="image" src="https://github.com/user-attachments/assets/aaa447ce-1fc9-4145-b0d2-52bc263d2a42" />

lukmanulhakkeem commented:
Microsoft supports multiple authentication mechanisms, including NTLM, and has provided documentation for each 

<img width="438" alt="image" src="https://github.com/user-attachments/assets/fea23373-1b30-45d1-9f34-55eb26ca26d9" />

https://learn.microsoft.com/en-us/sql/connect/jdbc/using-kerberos-integrated-authentication-to-connect-to-sql-server?view=sql-server-ver16

It would be difficult to include all of them in the Presto documentation. I don't think it’s ideal to document only NTLM authentication while leaving out other popular ones like Kerberos.

@imjalpreet @steveburnett Please suggest an approach here

steveburnett commented:
1. If the current documentation describes the default value of encrypt as ``true``, and that is how it will be after this PR is merged, then I think no doc change for this is needed. 

2. I agree that it would be awkward to mention only NTLM and not the other authentications, and that it is outside the scope of this PR to ask you to add documention for how to use all the other auth mechanisms that are supported. 

   I am okay with not mentioning authentication in the documentation in the SQL Server page, but I would ask that you consider mentioning NTLM as the motivation for the upgrade in a release note. Maybe something like this:
  
  ```
  SQL Server Connector Changes
  
  * Upgrade sqlserver driver to version 12.8.1, to include NTLM authentication support. 
  ```

   This way, _NTLM_ is at least in the Presto documentation as a searchable term for people to find if they search for it, and in the context of the SQL Server connector. 
    

3. Maybe we could add a heading _Authentication_ to the [SQL Server](https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/connector/sqlserver.rst) page. In it list the authentication methods that we know to work. Something as short as this: 
  
_Authentication_

_The following authentication methods are supported:_
_* NTLM_
_* Kerberos_
_* (others)_

   If you feel comfortable with doing this third ask in this PR, please do. You don't have to be complete, just list the ones you know to work successfully. The doc can be iterated on and expanded later. 

   If you do not feel comfortable with doing this third ask in this PR, I can open a new issue about adding an Authentication topic to the SQL Server page and the issue can be addressed in a later PR. In that case I would welcome your input as a contributor and reviewer.

lukmanulhakkeem commented:
@steveburnett I added authentication properties in brief. Could you please review

steveburnett reviewed (approved):
Thanks for this! I think this is perfect. 

I really appreciate you providing an example of a sample connection URL with the NTLM authentication. 

Other authentication method examples can be added in future documentation PRs, and this new Authentication topic gives a place for that new doc to be added.

steveburnett commented:
> Should I edit any files to add it to the release note?

I would suggest that you link to the new documentation from the release note. I tested this link (linking to a page is easy, linking to a heading within a page is a little trickier) and confirmed that it works in a local doc build. 

```
== RELEASE NOTES ==

SQL Server Connector Changes
* Upgrade sqlserver driver to version 12.8.1, to include NTLM authentication support. See :ref:`connector/sqlserver:authentication`.
```

imjalpreet commented:
> This default value is already updated in the documentation by mistake.

@lukmanulhakkeem I believe we should add a note stating that the default value of `encrypt` changed from `false` to `true` starting in Presto 0.292. Some users may not explicitly set `encrypt` to `false`, assuming the default was still `false`, just as we didn’t set it to `false` in our Product Tests.

lukmanulhakkeem commented:
@imjalpreet Can i add the below note?

`Starting from release 0.292, the default value of encrypt has changed from false to true.`

imjalpreet commented:
@lukmanulhakkeem Yes, looks good. @steveburnett can also review it once added.

You can use a note similar to the one used in this overview section: https://prestodb.github.io/docs/0.291/connector/druid.html.

lukmanulhakkeem commented:
Updated the doc.

imjalpreet commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -62,6 +62,46 @@ A connection string using a truststore would be similar to the following example
 
     connection-url=jdbc:sqlserver://<host>:<port>;databaseName=<databaseName>;encrypt=true;trustServerCertificate=false;trustStoreType=PEM;hostNameInCertificate=hostname;trustStore=path/to/truststore.pem;trustStorePassword=password
 
+.. note::
```
nit: let's move this to above the current documentation of the `encrypt` property.

imjalpreet replied:
Above this line: https://github.com/prestodb/presto/blob/7fb09c9fec2312f1252cd98f7bef4ed1ca83d793/presto-docs/src/main/sphinx/connector/sqlserver.rst?plain=1#L32

lukmanulhakkeem replied:
Okay

lukmanulhakkeem replied:
Done

steveburnett replied:
> Done

Could you check this move of text? I'm not seeing it in the current branch.

lukmanulhakkeem replied:
@steveburnett You can check this - https://github.com/prestodb/presto/compare/7fb09c9fec2312f1252cd98f7bef4ed1ca83d793..2354bc0c1a63a14e94fb3bdd83a7f1255024c246

imjalpreet reviewed (commented):
@lukmanulhakkeem just a small nit, otherwise LGTM, can you please squash your commits too? Thanks.

lukmanulhakkeem commented:
Extra space added with the commit. I am fixing

steveburnett reviewed (approved):
LGTM! (docs)

Pull updated branch, new local doc build, looks good.

lukmanulhakkeem commented:
Thanks @steveburnett

lukmanulhakkeem commented:
@imjalpreet Anything pending on this? Are we good to merge?

imjalpreet reviewed (commented):
@lukmanulhakkeem We cannot proceed with the merge until all the tests pass. It seems that one of the pipelines is failing due to some flaky tests. Could you please rebase on the latest master and push your changes again? I noticed that a few fixes for flaky tests have been added in the past week.

lukmanulhakkeem commented:
@imjalpreet All checks have passed now.

steveburnett commented:
> @lukmanulhakkeem We cannot proceed with the merge until all the tests pass. It seems that one of the pipelines is failing due to some flaky tests. Could you please rebase on the latest master and push your changes again? I noticed that a few fixes for flaky tests have been added in the past week.

@imjalpreet, all tests are passing on this PR now. When you have time, would you re-review and see if your earlier comments have been addressed?

lukmanulhakkeem commented:
@jaystarshot could you help merge this PR?

jaystarshot commented on presto-docs/src/main/sphinx/connector/sqlserver.rst:
```diff
@@ -29,6 +29,10 @@ Connection security
 The JDBC driver and connector automatically use Transport Layer Security (TLS) encryption and certificate validation. This requires a suitable TLS certificate configured on your SQL Server database host.
 
+.. note::
+
```
This seems like a breaking change to me. We should explicitly call this out in the release notes with examples.

lukmanulhakkeem replied:
@jaystarshot Is the below content good for the release note?

```
== RELEASE NOTES ==

SQL Server Connector Changes
* Upgraded SQL Server driver to version 12.8.1 to support NTLM authentication. See :ref:connector/sqlserver:authentication for more details.
* Note: Starting from this version, the driver sets the encrypt property to true by default. If you are connecting to a non-SSL SQL Server instance, you must explicitly set encrypt=false in your connection configuration to avoid connectivity issues.
```

jaystarshot replied:
Maybe add breaking change in the note somewhere

lukmanulhakkeem commented:
@jaystarshot Is the below content good for the release note?

```
== RELEASE NOTES ==

SQL Server Connector Changes
* Upgraded SQL Server driver to version 12.8.1 to support NTLM authentication. See :ref:connector/sqlserver:authentication for more details.
* Note: Starting from this version, the driver sets the encrypt property to true by default. If you are connecting to a non-SSL SQL Server instance, you must explicitly set encrypt=false in your connection configuration to avoid connectivity issues.
```

lukmanulhakkeem commented:
Thanks @jaystarshot . I added below content to the release note. Added the keyword `breaking change`

```
== RELEASE NOTES ==

SQL Server Connector Changes
* Upgraded SQL Server driver to version 12.8.1 to support NTLM authentication. See :ref:connector/sqlserver:authentication for more details.
* Note: Starting from this version, the driver sets the encrypt property to true by default. If you are connecting to a non-SSL SQL Server instance, you must explicitly set encrypt=false in your connection configuration to avoid connectivity issues. This is a breaking change for existing connections.

```
